src/main/java/hudson/maven/AbstractMavenProcessFactory.java

```diff
@@ -302,7 +302,11 @@ public String call() throws IOException {
                 if (originalJdk == null) { // so we only try this once
                     for (Throwable t = x; t != null; t = t.getCause()) {
                         if (t instanceof UnsupportedClassVersionError || t instanceof SocketException) {
-                            listener.error("[JENKINS-18403] JDK 5 not supported to run Maven; retrying with slave Java and setting compile/test properties to point to " + jdk.getHome());
+                            listener.error("[JENKINS-18403][JENKINS-28294] JDK '" + jdk.getName() +
+                                    "' not supported to run Maven projects; retrying with slave Java and setting compile/test properties to point to " +
+                                    jdk.getHome() +
+                                    ". Maven projects have to be launched with a Java version greater or equal to the one used by the slave. " +
```

Comment:
I did that test : Jenkins core 1.642.1 on JDK7 with Slave on JDK7 with Maven Job on JDK 6 launching a build with Maven 3.0.5 and it gives: 

------
02/05/16 00:14:51] [SSH] Starting sftp client.
[02/05/16 00:14:51] [SSH] Copying latest slave.jar...
[02/05/16 00:14:52] [SSH] Copied 506,256 bytes.
Expanded the channel window size to 4MB
[02/05/16 00:14:52] [SSH] Starting slave process: cd "/Users/arnaud/Temp/slave" && /Library/Java/JavaVirtualMachines/jdk1.7.0_80.jdk/Contents/Home/bin/java  -jar slave.jar
<===[JENKINS REMOTING CAPACITY]===>channel started
Slave.jar version: 2.53.2
------


------
[appaloosa] $ /Library/Java/JavaVirtualMachines/1.6.0.jdk/Contents/Home/bin/java -cp /Users/arnaud/Temp/slave/maven3-agent.jar:/Users/arnaud/Temp/slave/tools/hudson.tasks.Maven_MavenInstallation/maven-3.0.5/boot/plexus-classworlds-2.4.jar org.jvnet.hudson.maven3.agent.Maven3Main /Users/arnaud/Temp/slave/tools/hudson.tasks.Maven_MavenInstallation/maven-3.0.5 /Users/arnaud/Temp/slave/slave.jar /Users/arnaud/Temp/slave/maven3-interceptor.jar /Users/arnaud/Temp/slave/maven3-interceptor-commons.jar 60564
<===[JENKINS REMOTING CAPACITY]===>channel started
ERROR: [JENKINS-18403][JENKINS-28294] JDK 'java-1.6' not supported to run Maven projects; retrying with slave Java and setting compile/test properties to point to /Library/Java/JavaVirtualMachines/1.6.0.jdk/Contents/Home. Maven projects have to be launched with a JDK version greater or equal to the one used by the slave. Use the Maven JDK Toolchains (plugin) to build your maven project with an older JDK.
Established TCP socket on 60569
maven3-agent.jar already up to date
maven3-interceptor.jar already up to date
maven3-interceptor-commons.jar already up to date
[appaloosa] $ /Library/Java/JavaVirtualMachines/jdk1.7.0_80.jdk/Contents/Home/jre/bin/java -cp /Users/arnaud/Temp/slave/maven3-agent.jar:/Users/arnaud/Temp/slave/tools/hudson.tasks.Maven_MavenInstallation/maven-3.0.5/boot/plexus-classworlds-2.4.jar org.jvnet.hudson.maven3.agent.Maven3Main /Users/arnaud/Temp/slave/tools/hudson.tasks.Maven_MavenInstallation/maven-3.0.5 /Users/arnaud/Temp/slave/slave.jar /Users/arnaud/Temp/slave/maven3-interceptor.jar /Users/arnaud/Temp/slave/maven3-interceptor-commons.jar 60569
<===[JENKINS REMOTING CAPACITY]===>channel started
------

Maven 3.0.5 supports Java 6 but we can see the error when it is launched with this version. It is then started with Java 7. AFAIU the problem comes from the fact that some J7 serialised classes coming from master are executing on the J6.